### PR TITLE
Pil 281: Refactoring  of Registration and Subscription 

### PR DESCRIPTION
**Pull request by alikouhestani**

### CODE REVIEW

File: app/connectors/EnrolmentConnector.scala [hunk 1]
```diff
@@ -16,36 +16,31 @@
...
-) extends Logging {
-
-  def createEnrolment(
-    enrolmentInfo: EnrolmentInfo
-  )(implicit hc:   HeaderCarrier, ec: ExecutionContext): Future[Option[Int]] = {
-
-    val url: String = s"${config.taxEnrolmentsUrl1}HMRC-PILLAR2-ORG${config.taxEnrolmentsUrl2}"
+@Singleton
+class EnrolmentConnector @Inject() (val config: FrontendAppConfig, val http: HttpClient)(implicit ec: ExecutionContext) extends Logging {
+  val url: String = s"${config.taxEnrolmentsUrl1}HMRC-PILLAR2-ORG${config.taxEnrolmentsUrl2}"
```
Comment (jassalrichy) [inline app/connectors/EnrolmentConnector.scala:33]: get enrolment.key from application.conf

File: app/connectors/EnrolmentConnector.scala [hunk 2]
```diff
@@ -16,36 +16,31 @@
...
 
-    http.PUT[EnrolmentRequest, HttpResponse](url, enrolmentInfo.convertToEnrolmentRequest) map {
-      case responseMessage if is2xx(responseMessage.status) =>
-        Some(responseMessage.status)
-      case responseMessage =>
+  def createEnrolment(enrolmentInfo: EnrolmentInfo)(implicit hc: HeaderCarrier): Future[Done] =
+    http.PUT[EnrolmentRequest, HttpResponse](url, enrolmentInfo.convertToEnrolmentRequest) flatMap {
+      case success if is2xx(success.status) => Done.toFuture
```
Comment (jassalrichy) [inline app/connectors/EnrolmentConnector.scala:37]: General: Should we deal with the response within the Controller/Service rather than within the Connectors ?

File: app/connectors/UserAnswersConnectors.scala [hunk 3]
```diff
@@ -51,11 +50,11 @@ class UserAnswersConnectors @Inject() (
     }
 
   def getUserAnswer(id: String)(implicit headerCarrier: HeaderCarrier): Future[Option[UserAnswers]] =
```
Comment (jassalrichy) [inline app/connectors/UserAnswersConnectors.scala:52]: Need unit tests for getUserAnswer

File: app/controllers/CheckYourAnswersController.scala [hunk 4]
```diff
@@ -114,19 +115,39 @@ class CheckYourAnswersController @Inject() (
...
+              case InternalIssueError =>
+                logger.error(s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Subscription failed due to failed call to the backend")
+                Redirect(controllers.subscription.routes.SubscriptionFailedController.onPageLoad)
+
+              case DuplicateSubmissionError =>
+                logger.error(s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Subscription failed due to a Duplicate Submission")
+                Redirect(controllers.routes.AlreadyRegisteredController.onPageLoad)
+            }
+        }
+        .getOrElse(Future.successful(Redirect(controllers.routes.JourneyRecoveryController.onPageLoad())))
```
Comment (jassalrichy) [inline app/controllers/CheckYourAnswersController.scala:148]: add test for redirect to journey recovery controller

File: app/controllers/DashboardController.scala [hunk 5]
```diff
@@ -44,88 +44,64 @@ class DashboardController @Inject() (
...
+          } yield optionalUserAnswers match {
+            case Some(userAnswers) =>
+              userAnswers
+                .get(fmDashboardPage)
+                .map { dashboard =>
+                  val inactiveStatus = request.userAnswers
                     .get(subAccountStatusPage)
-                    .map { acctStatus =>
+                    .exists { acctStatus =>
                       acctStatus.inactive
```
Comment (jassalrichy) [inline app/controllers/DashboardController.scala:79]: Add test to return acctStatus.inactive

File: app/controllers/DashboardController.scala [hunk 6]
```diff
@@ -44,88 +44,64 @@ class DashboardController @Inject() (
...
-                s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Duplicate submission detected."
-              case UnprocessableEntityError =>
-                s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Unprocessable entity error."
-              case InternalServerError_ =>
-                s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Internal server error."
-              case ServiceUnavailableError =>
-                s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Service Unavailable error."
-              case SubscriptionCreateError =>
-                s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Subscription creation error."
+            case _ => Redirect(controllers.routes.JourneyRecoveryController.onPageLoad())
```
Comment (jassalrichy) [inline app/controllers/DashboardController.scala:93]: add test for redirect to journey recovery controller

File: app/controllers/MakeAPaymentDashboardController.scala [hunk 7]
```diff
@@ -18,40 +18,39 @@ package controllers
...
-
+  def onPageLoad: Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    sessionRepository
+      .get(request.userId)
+      .map { optionalUserAnswers =>
+        (for {
+          userAnswers <- optionalUserAnswers
+          pillar2Id <- Pillar2Reference
+                         .getPillar2ID(request.enrolments, appConfig.enrolmentKey, appConfig.enrolmentIdentifier)
+                         .orElse(userAnswers.get(plrReferencePage))
```
Comment (jassalrichy) [inline app/controllers/MakeAPaymentDashboardController.scala:51]: Add test to retrieve pillar2Id from userAnswers plrReferencePage

File: app/controllers/RegistrationConfirmationController.scala [hunk 8]
```diff
@@ -34,19 +35,24 @@ class RegistrationConfirmationController @Inject() (
...
     val currentDate = HtmlFormat.escape(dateHelper.formatDateGDS(java.time.LocalDate.now))
-    val mneDomestic = request.session.data.get(Pillar2SessionKeys.updateMneOrDomestic) match {
-      case Some("ukAndOther") => appConfig.registrationControllerMne
-      case _                  => appConfig.registrationControllerDomestic
+    sessionRepository.get(request.userAnswers.id).map { optionalUserAnswers =>
+      (for {
+        userAnswer <- optionalUserAnswers
+        pillar2Id <- Pillar2Reference
+                       .getPillar2ID(request.enrolments, appConfig.enrolmentKey, appConfig.enrolmentIdentifier)
+                       .orElse(userAnswer.get(plrReferencePage))
```
Comment (jassalrichy) [inline app/controllers/RegistrationConfirmationController.scala:52]: Add test to retrieve pillar2Id from userAnswers plrReferencePage

File: app/controllers/TaskListController.scala [hunk 9]
```diff
@@ -136,21 +139,18 @@ class TaskListController @Inject() (
...
-        Future.successful(
-          Ok(
-            view(ultimateParentInfo, count, filingMemberInfo, groupDetailInfo, contactDetailsInfo, cyaInfo)
-          )
-        )
+    sessionRepository.get(request.userId).flatMap { optionalUA =>
+      optionalUA.map(userAnswers => userAnswers.get(plrReferencePage).isDefined) match {
+
+        case Some(true) => Future.successful(Redirect(routes.RegistrationConfirmationController.onPageLoad))
+        case _ if pillar2ReferenceFromReadSubscription =>
```
Comment (jassalrichy) [inline app/controllers/TaskListController.scala:146]: Add test for when pillar2ReferenceFromReadSubscription is true and no value exists for userAnswers plrReferencePage

File: app/controllers/eligibility/BusinessActivityUKController.scala [hunk 10]
```diff
@@ -16,55 +16,69 @@
...
-    val preparedForm = request.session.data.get(Pillar2SessionKeys.businessActivityUKPageYesNo) match {
-      case None        => form
-      case Some(value) => form.fill(value)
+  def onPageLoad: Action[AnyContent] = Action.async { implicit request =>
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).map { OptionalUserAnswers =>
+      val preparedForm = OptionalUserAnswers.getOrElse(UserAnswers(sessionID)).get(BusinessActivityUKPage) match {
+        case None       => form
+        case Some(data) => form.fill(data)
```
Comment (jassalrichy) [inline app/controllers/eligibility/BusinessActivityUKController.scala:52]: Add test for when question has already been answered on page load.

File: app/controllers/eligibility/GroupTerritoriesController.scala [hunk 11]
```diff
@@ -16,57 +16,68 @@
...
-    val preparedForm = request.session.data.get(Pillar2SessionKeys.groupTerritoriesPageYesNo) match {
-      case None        => form
-      case Some(value) => form.fill(value)
+  def onPageLoad: Action[AnyContent] = Action.async { implicit request =>
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).map { OptionalUserAnswers =>
+      val preparedForm = OptionalUserAnswers.getOrElse(UserAnswers(sessionID)).get(UpeEqPage) match {
+        case None       => form
+        case Some(data) => form.fill(data)
```
Comment (jassalrichy) [inline app/controllers/eligibility/GroupTerritoriesController.scala:52]: Add test for when question has already been answered on page load.

File: app/controllers/eligibility/RegisteringNfmForThisGroupController.scala [hunk 12]
```diff
@@ -32,57 +32,69 @@
...
-    val preparedForm = request.session.data.get(Pillar2SessionKeys.registeringNfmForThisGroup) match {
-      case None        => form
-      case Some(value) => form.fill(value)
+  def onPageLoad: Action[AnyContent] = Action.async { implicit request =>
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).map { OptionalUserAnswers =>
+      val preparedForm = OptionalUserAnswers.getOrElse(UserAnswers(sessionID)).get(NfmEqPage) match {
+        case None       => form
+        case Some(data) => form.fill(data)
```
Comment (jassalrichy) [inline app/controllers/eligibility/RegisteringNfmForThisGroupController.scala:68]: Add test for when question has already been answered on page load.

File: app/services/SubscriptionService.scala [hunk 13]
```diff
@@ -16,50 +16,48 @@
...
+  registrationConnector:        RegistrationConnector,
+  subscriptionConnector:        SubscriptionConnector,
+  enrolmentConnector:           EnrolmentConnector,
+  enrolmentStoreProxyConnector: EnrolmentStoreProxyConnector
+)(implicit ec:                  ExecutionContext) {
+
+  def createSubscription(userAnswers: UserAnswers)(implicit hc: HeaderCarrier): Future[String] =
+    for {
+      upeSafeId       <- registerUpe(userAnswers)
+      fmnSafeId       <- registerNfm(userAnswers)
```
Comment (jassalrichy) [inline app/services/SubscriptionService.scala:40]: Add test which return none when calling registerNfm

File: app/connectors/UserAnswersConnectors.scala (hunk 3 above)
Comment (alikouhestani) [reply app/connectors/UserAnswersConnectors.scala:52]: Done (except for the 200 response case)

File: app/controllers/DashboardController.scala (hunk 6 above)
Comment (alikouhestani) [reply app/controllers/DashboardController.scala:93]: Done

File: app/controllers/DashboardController.scala (hunk 5 above)
Comment (alikouhestani) [reply app/controllers/DashboardController.scala:79]: amended one of the test cases so that it covers this scenario

File: app/controllers/CheckYourAnswersController.scala (hunk 4 above)
Comment (alikouhestani) [reply app/controllers/CheckYourAnswersController.scala:148]: that case will never happen as the "finalStatusCheck" only returns true if subMneOrDomesticPage value exists. Happy to discuss this

File: app/connectors/EnrolmentConnector.scala (hunk 2 above)
Comment (alikouhestani) [reply app/connectors/EnrolmentConnector.scala:37]: I was under the impression that we (me, you and @AKKUM ) agreed on this design. The response is dealt with in the controller, its just that here we raise a Future.failed if there is anything wrong with the http request and deal with the error in the controller level. Happy to discuss this

File: app/controllers/RegistrationConfirmationController.scala (hunk 8 above)
Comment (alikouhestani) [reply app/controllers/RegistrationConfirmationController.scala:52]: Done

File: app/controllers/eligibility/BusinessActivityUKController.scala (hunk 10 above)
Comment (alikouhestani) [reply app/controllers/eligibility/BusinessActivityUKController.scala:52]: Done

File: app/controllers/eligibility/GroupTerritoriesController.scala (hunk 11 above)
Comment (alikouhestani) [reply app/controllers/eligibility/GroupTerritoriesController.scala:52]: Done

File: app/controllers/eligibility/RegisteringNfmForThisGroupController.scala (hunk 12 above)
Comment (alikouhestani) [reply app/controllers/eligibility/RegisteringNfmForThisGroupController.scala:68]: Done

File: app/controllers/eligibility/TurnOverEligibilityController.scala [hunk 14]
```diff
@@ -16,56 +16,68 @@
...
-    val preparedForm = request.session.data.get(Pillar2SessionKeys.turnOverEligibilityValue) match {
-      case None        => form
-      case Some(value) => form.fill(value)
+  def onPageLoad: Action[AnyContent] = Action.async { implicit request =>
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).map { OptionalUserAnswers =>
+      val preparedForm = OptionalUserAnswers.getOrElse(UserAnswers(sessionID)).get(RevenueEqPage) match {
+        case None       => form
+        case Some(data) => form.fill(data)
```
[the remark this answers is not in the document]
Comment (alikouhestani) [reply app/controllers/eligibility/TurnOverEligibilityController.scala:52]: Done

File: app/controllers/MakeAPaymentDashboardController.scala (hunk 7 above)
Comment (alikouhestani) [reply app/controllers/MakeAPaymentDashboardController.scala:51]: Done

File: app/controllers/TaskListController.scala (hunk 9 above)
Comment (alikouhestani) [reply app/controllers/TaskListController.scala:146]: Done

**Review by jassalrichy — DISMISSED** [review]

File: app/connectors/SubscriptionConnector.scala [hunk 15]
```diff
@@ -17,20 +17,33 @@
...
-    subscriptionParameter: SubscriptionRequestParameters
-  )(implicit hc:           HeaderCarrier, ec: ExecutionContext): Future[HttpResponse] =
-    http.POST[SubscriptionRequestParameters, HttpResponse](subscriptionUrl, subscriptionParameter)
+@Singleton
+class SubscriptionConnector @Inject() (val config: FrontendAppConfig, val http: HttpClient)(implicit ec: ExecutionContext) extends Logging {
+  val subscriptionUrl = s"${config.pillar2BaseUrl}/report-pillar2-top-up-taxes/subscription/create-subscription"
 
+  def subscribe(subscriptionRequestParameters: SubscriptionRequestParameters)(implicit hc: HeaderCarrier): Future[String] =
+    http
+      .POST[SubscriptionRequestParameters, HttpResponse](s"$subscriptionUrl", subscriptionRequestParameters)
```
Comment (jassalrichy) [inline app/connectors/SubscriptionConnector.scala:39]: String interpolation not require, just pass (subscriptionUrl, subscriptionRequestParameters)

File: app/controllers/TaskListController.scala [hunk 16]
```diff
@@ -136,21 +139,18 @@ class TaskListController @Inject() (
...
+          userAnswersConnectors.remove(request.userId).map { _ =>
+            logger.info(s"Remove existing amend data from local database if exist")
```
Comment (jassalrichy) [inline app/controllers/TaskListController.scala:148]: can we prefix this logger.info message with "**[Session ID: ${Pillar2SessionKeys.sessionId(hc)}]** - ..."

File: app/controllers/eligibility/GroupTerritoriesController.scala [hunk 17]
```diff
@@ -16,57 +16,68 @@
...
-      )
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).flatMap { optionalUserAnswer =>
+      val userAnswer = optionalUserAnswer.getOrElse(UserAnswers(sessionID))
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(view(formWithErrors))),
+          value =>
```
Comment (jassalrichy) [inline app/controllers/eligibility/GroupTerritoriesController.scala:67]: Converted match statement to anonymous function (see BusinessActivityUKController above)

File: app/controllers/eligibility/RegisteringNfmForThisGroupController.scala [hunk 18]
```diff
@@ -32,57 +32,69 @@
...
-      )
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
+    sessionRepository.get(sessionID).flatMap { optionalUserAnswer =>
+      val userAnswer = optionalUserAnswer.getOrElse(UserAnswers(sessionID))
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(view(formWithErrors))),
+          value =>
```
Comment (jassalrichy) [inline app/controllers/eligibility/RegisteringNfmForThisGroupController.scala:83]: Converted match statement to anonymous function (see BusinessActivityUKController above)

File: app/services/ReadSubscriptionService.scala [hunk 19]
```diff
@@ -29,38 +28,12 @@ import scala.concurrent.{ExecutionContext, Future}
...
-      case 404 => NotFoundError
-      case 409 => DuplicateSubmissionError
-      case 422 => UnprocessableEntityError
-      case 500 => InternalServerError_
-      case 503 => ServiceUnavailableError
+        Future.successful(jsValue)
       case _ =>
-        logger.error(s"Unhandled error code: $errorCode with detail: ${Json.stringify(errorDetail)}")
-        InternalServerError_
+        Future.failed(InternalIssueError)
```
Comment (jassalrichy) [inline app/services/ReadSubscriptionService.scala:37]: Does this logic work with the stub for replicating BadRequestError, NotFoundError etc ?

File: app/controllers/eligibility/GroupTerritoriesController.scala (hunk 17 above)
Comment (alikouhestani) [reply app/controllers/eligibility/GroupTerritoriesController.scala:67]: Happy to do so but what are we achieving by writing the match statement that way?

File: app/controllers/eligibility/RegisteringNfmForThisGroupController.scala (hunk 18 above)
Comment (alikouhestani) [reply app/controllers/eligibility/RegisteringNfmForThisGroupController.scala:83]: Again, similarly, what is achieved by writing it as "value @true"  as opposed to "case true"

File: app/services/ReadSubscriptionService.scala (hunk 19 above)
Comment (alikouhestani) [reply app/services/ReadSubscriptionService.scala:37]: All acceptance tests are all passing so I feel like it is safe to assume that this branch and main offer the same functionalities in that regard

File: app/controllers/TaskListController.scala (hunk 16 above)
Comment (alikouhestani) [reply app/controllers/TaskListController.scala:148]: Good idea, thanks. Done

File: app/controllers/eligibility/RegisteringNfmForThisGroupController.scala (hunk 18 above)
Comment (olakunlekuye2) [reply app/controllers/eligibility/RegisteringNfmForThisGroupController.scala:83]: It is said to ensure the compiler applies optimisation to the pattern match expression

File: app/controllers/eligibility/BusinessActivityUKController.scala [hunk 20]
```diff
@@ -16,55 +16,69 @@
...
 
-  val form = formProvider()
+  val form: Form[Boolean] = formProvider()
 
-  def onPageLoad: Action[AnyContent] = Action { implicit request =>
-    val preparedForm = request.session.data.get(Pillar2SessionKeys.businessActivityUKPageYesNo) match {
-      case None        => form
-      case Some(value) => form.fill(value)
+  def onPageLoad: Action[AnyContent] = Action.async { implicit request =>
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
```
Comment (TevynAllen) [inline app/controllers/eligibility/BusinessActivityUKController.scala:47]: this should implicitly come from FrontendBaseController?

File: app/controllers/eligibility/BusinessActivityUKController.scala [hunk 21]
```diff
@@ -16,55 +16,69 @@
...
-                Redirect(controllers.eligibility.routes.TurnOverEligibilityController.onPageLoad)
-                  .withSession((sessionData.updateBusinessActivityUKYesNo(value)))
-              )
-            case "no" =>
-              Future.successful(
-                Redirect(controllers.eligibility.routes.KbUKIneligibleController.onPageLoad)
-                  .withSession((sessionData.updateBusinessActivityUKYesNo(value)))
-              )
+    val hc        = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val sessionID = Pillar2SessionKeys.sessionId(hc)
```
Comment (TevynAllen) [inline app/controllers/eligibility/BusinessActivityUKController.scala:60]: could probably create a nice way of implicitly having this in scope, similar to the comment above we can create our own BaseController that has HeaderCarrier and all the IDs we require

File: app/controllers/eligibility/BusinessActivityUKController.scala [hunk 22]
```diff
@@ -16,55 +16,69 @@
...
+    sessionRepository.get(sessionID).flatMap { optionalUserAnswer =>
```
Comment (TevynAllen) [inline app/controllers/eligibility/BusinessActivityUKController.scala:61]: As it stands we're making a call to the database if there is no session ID we should avoid this i.e. with `No Session ID available`. Maybe we can avoid this by letting the session repository accept a Option

File: app/views/BusinessActivityUKView.scala.html [hunk 23]
```diff
@@ -37,10 +37,9 @@
         }
         <span class="govuk-caption-l">@messages("businessActivityUK.check.liable.text")</span>
         @govukRadios(
-            RadiosViewModel(
+                RadiosViewModel.yesNo(
```
Comment (TevynAllen) [inline app/views/BusinessActivityUKView.scala.html:40]: 👍

File: app/views/EntityTypeView.scala.html [hunk 24]
```diff
@@ -45,7 +45,7 @@
             )
         )
         <p class="govuk-body">
-            <a class="govuk-link" href="/report-pillar2-top-up-taxes/business-matching/ultimate-parent/no-id/input-name">
+            <a class="govuk-link"  href="@controllers.registration.routes.UpeNameRegistrationController.onPageLoad(NormalMode)">
```
Comment (TevynAllen) [inline app/views/EntityTypeView.scala.html:48]: 👍

File: app/controllers/eligibility/BusinessActivityUKController.scala (hunk 22 above)
Comment (alikouhestani) [reply app/controllers/eligibility/BusinessActivityUKController.scala:60]: Lets have a chat about this tomorrow

File: app/controllers/eligibility/GroupTerritoriesController.scala (hunk 17 above)
Comment (jassalrichy) [reply app/controllers/eligibility/GroupTerritoriesController.scala:65]: Code will look a little more pleasant and also more readable.

A pattern matching anonymous function is an anonymous function that is defined as a block consisting of a sequence of cases, surrounded as usual by curly braces, but without a match keyword before the block.

File: app/controllers/eligibility/BusinessActivityUKController.scala (hunk 20 above)
Comment (alikouhestani) [reply app/controllers/eligibility/BusinessActivityUKController.scala]: Done

**Review by TevynAllen — APPROVED** [review]

**Review by jassalrichy — APPROVED** [review]
